### ISSUE TITLE
concrete `cartesian_product(::Zonotope, ::Zonotope)`

### ISSUE THREAD

**Issue by goretkin** [JuliaReach/LazySets.jl#2679]
Something like

```julia
function LazySets.cartesian_product(Z1::Zonotope, Z2::Zonotope)
    center = vcat(Z1.center, Z2.center)
    z1 = zeros(dim(Z1), size(Z2.generators, 2))
    z2 = zeros(dim(Z2), size(Z1.generators, 2))
    g1 = Z1.generators
    g2 = Z2.generators
    generators = [
        g1 z1;
        g2 z2
    ]
    return Zonotope(center, generators)
end
```

I wasn't sure if this was an omission, or intentionally unimplemented

**Comment by schillic** [comment]
> I wasn't sure if this was an omission, or intentionally unimplemented

There are probably many combinations that we do not support yet. Typically that means we never needed them in practice.

Thanks, your implementation makes sense to me (with this small fix).
```julia
...
    generators = [
        g1 z1;
        z2 g2
    ]
...
```
I wonder if using a special structure for this block matrix with 50% zeros would make sense.

The code can also be generalized to `AbstractZonotope` by using the functions `center`/`generators`/`ngens`, but practically the only interesting subtype is `Zonotope`.

```julia
function cartesian_product2(Z1::AbstractZonotope, Z2::AbstractZonotope)
    c = vcat(center(Z1), center(Z2))
    z1 = zeros(dim(Z1), ngens(Z2))
    z2 = zeros(dim(Z2), ngens(Z1))
    g1 = generators(Z1)
    g2 = generators(Z2)
    G = [
        g1 z1;
        z2 g2
    ]
    return Zonotope(c, G)
end
```

Do you want to send a PR?

**Comment by mforets** [comment]
Seems like a missing method. The idea you proposed lgtm.

There is a fallback, but it is suboptimal since it passes to the hrep:

```julia
julia> using LazySets

julia> Z = rand(Zonotope, dim=2);

julia> Y = rand(Zonotope, dim=2);

julia> cartesian_product(Z, Y)
ERROR: AssertionError: package 'Polyhedra' not loaded (it is required for executing ``cartesian_product`)
.....

julia> using Polyhedra

julia> cartesian_product(Z, Y)
HPolyhedron{Float64, Vector{Float64}}(LazySets.HalfSpace{Float64, Vector{Float64}}[LazySets.HalfSpace .....
```

**Comment by mforets** [comment]
@schillic  you beat me by 22 seconds :laughing:

**Comment by mforets** [comment]
BTW adding the method at the `AbstractZonotope` level would also help a bit with other cases such as hyperrectangles:

```julia
julia> H = rand(Hyperrectangle); Q = rand(Hyperrectangle)
Hyperrectangle{Float64, Vector{Float64}, Vector{Float64}}([0.8030341172908043, 0.010173278885805047], [0.22679035382167936, 0.4899978455677334])

julia> cartesian_product(H, Q)
HPolyhedron{Float64, Vector{Float64}}(LazySets.HalfSpace{Float64, Vector{Float64}}[LazySets.HalfSpace{Float64, Vector{Float64}}([1.0, 0.0, 0.0, 0.0], 1.2142622097124542), LazySets.HalfSpace{Float64, Vector{Float64}}([0.0, 1.0, 0.0, 0.0], 0.58122386849
```
Clearly adding an efficient method for `AbstractHyperrectangle` would be a different issue.

**Comment by goretkin** [comment]
> I wonder if using a special structure for this block matrix with 50% zeros would make sense.

I asked a [question on zulip](https://julialang.zulipchat.com/#narrow/stream/228422-linear-algebra/topic/construct.20block.20matrices.20without.20much.20thought) and was led to https://github.com/invenia/BlockDiagonals.jl

My guess is that for "small" matrices, it would be slower due to the non-uniform memory layout. But for "large" matrices, it would be a win. And if the matrices are already sparse, then they should probably be combined into a new sparse matrix? The sparse case could be easy to handle, and the result would be type stable.

Thanks for the more generic examples! I'll consider making a PR with the simple case. Perhaps someone can comment on the viability of taking on a dependency like `BlockDiagonals.jl`

**Comment by mforets** [comment]
Making the concatenation through `BlockDiagonals` is OK but seems like an overkill to use this in the library.

Usually it is enough to address the following 3 implementations: static arrays case (efficient for low dims); sparse case (same as above but using `spzeros`), and a fallback case for any `AbstractMatrix`, that implements the method described above (ie assuming the matrices were dense and converting if they arent') described above and that returns regular (dense) array.

**Comment by mforets** [comment]
To move forward this issue may just consider the method from above (fallback for any matrix type); the other specialized cases can be added in separate issues.. Divide and conquer.. :)

**Comment by mforets** [comment]
To extend a bit on the implementation, a pattern that we regularly use is to write two functions, one that works at the interface level (in this case, a pair of `AbstractZonotope`) that calls the actual implementation in another method at the concrete set type level (or an internal, unexported function, depending on the context).

In this case it could be:

```julia
function cartesian_product(X::AbstractZonotope, Y::AbstractZonotope)
    Z1 = convert(X, Zonotope)
    Z2 = convert(Y, Zonotope)
     return cartesian_product(Z1, Z2)
end

function cartesian_product(Z1::Zonotope, Z2::Zonotope)
     ...
end
```

Then in the future we can add new methods where the matrix type of `Z1` or `Z2` are of some specific matrix type.

**Comment by schillic** [comment]
What is the benefit over https://github.com/JuliaReach/LazySets.jl/issues/2679#issuecomment-827292828 ? I think that version will do exactly the same but without the additional `convert`.

**Comment by mforets** [comment]
I agree, the differences are only subtle and in some cases, one strategy or the other can help for generality, or to reduce code dup. What I had in mind is that if `X` is *not* a zonotope but its `genmat` returns, say, a static array, then we'd like that it also uses the implementation that returns the static array.

**Comment by schillic** [comment]
I went ahead and implemented it in #2679.
